keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,560 @@
+    volumeMounts:
+    - name: config
+      mountPath: /data
+    - name: data
+      mountPath: /etc/config
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: config.yaml
+          volumeName: config
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+  - name: data
+    hostPath: /etc/config
+```
+
+## Design Details
+
+Pods and PodTemplate will include `FileEnvSource` field under [`EnvFromSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvFromSource)
+and `FileKeySelector` field under [`EnvVarSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvVarSource)
+that can be set for an individual container.
+
+### Pod API
+
+`FileEnvSource` field will allow the users to populate environment variables
+from the specified file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+```
+type EnvFromSource struct {
+    ...
+    // The file to select from
+    // +optional
+    FileRef *FileEnvSource `json:"fileRef,omitempty" protobuf:"bytes,3,opt,name=fileRef"`
+    ...
+}
+
+type FileEnvSource struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // Specify whether the file must exist.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,3,opt,name=optional"`
+}
+```
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file.
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from.  Must be a valid key.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
+    // Specify whether the file or its key must be defined
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
+}
+```
+
+### Env File
+
+The full specification of an env file:
+
+1. **File Format**: The environment variable (env) file must adhere to valid YAML syntax to ensure correct parsing.
```

Comment:
Why yaml? I'm not sure it's a standard, but there's a convention for `*.env` files to be in a `KEY=VALUE` format, essentially something that can be loaded directly in bash with `source file.env`.

Reply:
Agree, Sergey also above :)

Reply:
changed this to .env